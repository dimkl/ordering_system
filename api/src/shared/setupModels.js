const { Model } = require('objection');
const knex = require('./knex');

module.exports = () => { Model.knex(knex) }