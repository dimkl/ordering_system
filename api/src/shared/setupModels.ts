import { Model } from "objection";
import { knex } from "./knex";

export default () => {
  Model.knex(knex);
  return knex;
};
