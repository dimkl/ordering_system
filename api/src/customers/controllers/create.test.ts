/**
 * @integration-test true
 * @data-factory true
 */

import type { Knex } from "knex";

import setupModels from "../../shared/setupModels";
import { ulid } from "ulid";

describe("POST /customers", () => {
  let knex: Knex;
  beforeAll(() => (knex = setupModels()));
  beforeEach(() => knex.raw("truncate customers, orders cascade;"));
  afterAll(() => knex.destroy());

  it("creates and returns a customer", async () => {
    const response = await request
      .post("/customers")
      .send({
        first_name: "Dimitris",
        last_name: "Klouvas",
        email: "dimitris.klouvas@gmail.com",
        password: "1234"
      })
      .set("Accept", "application/json");

    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot({
      id: expect.any(String),
      created_at: expect.any(String),
      updated_at: expect.any(String)
    });
  });

  it("throws validation error for required properties", async () => {
    const response = await request.post("/customers").send({}).set("Accept", "application/json");

    expect(response.status).toBe(400);
    expect(response.body).toMatchSnapshot();
  });

  it("throws validation error for unique email", async () => {
    const email = "customer@example.com";
    await DataFactory.createCustomer({
      first_name: "First",
      last_name: "Customer",
      email,
      password: "1234"
    });

    const response = await request
      .post("/customers")
      .send({
        first_name: "Dimitris",
        last_name: "Klouvas",
        email: email,
        password: "1234"
      })
      .set("Accept", "application/json");

    expect(response.status).toBe(422);
    expect(response.body).toMatchSnapshot();
  });

  it("omits additional properties", async () => {
    const customId = ulid();
    const response = await request
      .post("/customers")
      .send({
        first_name: "Dimitris",
        last_name: "Klouvas",
        email: "dimitris.klouvas@gmail.com",
        password: "1234",
        id: customId,
        created_at: "1680046371850"
      })
      .set("Accept", "application/json");

    expect(response.status).toBe(200);
    expect(response.body.id).not.toEqual(customId);
    expect(response.body.created_at).not.toEqual("1680046371850");
    expect(response.body).toMatchSnapshot({
      id: expect.any(String),
      created_at: expect.any(String),
      updated_at: expect.any(String)
    });
  });
});
