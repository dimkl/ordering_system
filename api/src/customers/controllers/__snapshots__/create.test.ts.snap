// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /customers creates and returns a customer 1`] = `
{
  "created_at": Any<String>,
  "email": "dimitris.klouvas@gmail.com",
  "first_name": "Dimitris",
  "id": Any<Number>,
  "last_name": "Klouvas",
  "updated_at": Any<String>,
  "uuid": Any<String>,
}
`;

exports[`POST /customers omits additional properties 1`] = `
{
  "created_at": Any<String>,
  "email": "dimitris.klouvas@gmail.com",
  "first_name": "Dimitris",
  "id": Any<Number>,
  "last_name": "Klouvas",
  "updated_at": Any<String>,
  "uuid": Any<String>,
}
`;

exports[`POST /customers throws validation error for required properties 1`] = `
[
  {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'first_name'",
    "params": {
      "missingProperty": "first_name",
    },
    "schemaPath": "#/required",
  },
  {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'last_name'",
    "params": {
      "missingProperty": "last_name",
    },
    "schemaPath": "#/required",
  },
  {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'email'",
    "params": {
      "missingProperty": "email",
    },
    "schemaPath": "#/required",
  },
  {
    "instancePath": "",
    "keyword": "required",
    "message": "must have required property 'password'",
    "params": {
      "missingProperty": "password",
    },
    "schemaPath": "#/required",
  },
]
`;

exports[`POST /customers throws validation error for unique email 1`] = `
{
  "message": "email already exists!",
}
`;
